Add a caching layer to pachyderm

jdoliner opened the pull request.

jdoliner commented:
@sjezewski if you wouldn't mind making a pass on this

sjezewski commented on src/server/pfs/server/obj_block_api_server.go:
```diff
@@ -155,29 +187,20 @@ func (s *objBlockAPIServer) PutBlock(putBlockServer pfsclient.BlockAPI_PutBlockS
+	} else {
+		data = nil
+	}
+	getBlockServer.Send(&google_protobuf.BytesValue{Value: data})
```
I think you just want to return this value. Otherwise if it errs, we don't report it

jdoliner replied:
Great catch, I'm not sure what I was thinking here

sjezewski commented on src/server/pkg/cache/server/server.go:
```diff
@@ -0,0 +1,114 @@
+	request *groupcachepb.GetRequest,
+) (response *groupcachepb.GetResponse, retErr error) {
+	// response is too big to log
+	defer func(start time.Time) { s.Log(request, nil, retErr, time.Since(start)) }(time.Now())
```
Do we want to follow the new paradigm where we also provide a log statement at the beginning of the function?

If I understand the get logic correctly, if its a cache miss the getter could try to read from obj storage and hang. So, it may be useful to instrument here as well w a log line at the beginning of the function.

jdoliner replied:
Yup will do.

sjezewski commented on src/server/pfs/server/obj_block_api_server.go:
```diff
@@ -39,6 +41,36 @@ func newObjBlockAPIServer(dir string, objClient obj.Client) (*objBlockAPIServer,
 		dir:         dir,
 		localServer: localServer,
 		objClient:   objClient,
+		cache: groupcache.NewGroup("block", 1024*1024*1024*10,
```
We should provide a way to enable/disable. How about (per your comments :-P) provide size as a config flag?

sjezewski commented:
And one last thing. Once we have a flag to enable/disable, this seems like a great example to write a benchmark to validate the behavior.

After that, LGTM

![image](https://cloud.githubusercontent.com/assets/597953/17792839/21a0792e-6558-11e6-8c49-ae7bd1d14dd2.png)

jdoliner commented:
K, I think I've addressed all of your comments except for the one about writing a benchmark. Basically writing a benchmark would require 2 things, one is writing a simple golang benchmark which reads and writes files. The other is having a way to deploy the server with and without a cache. I think it'd take a lot of scaffolding to get that working so I'd prefer to get this merged in and then do a bigger push on benchmarking and performance in the next release cycle when we'll write all of the scaffolding code and come up with different workloads. Does that seem reasonable?

sjezewski commented:
Yea, thats fair. We have an open issue ( #648 ) about having test coverage for an obj store ... once we have that in place, this gets a lot easier.

Let's make a new issue for this benchmark and refer to it in #648
